pkg/storegateway/indexheader/index/postings.go

```diff
@@ -0,0 +1,514 @@
+			prevRng.End = int64(off - crc32.Size)
+			t.postings[lastKey[0]][lastKey[1]] = prevRng
+		}
+
+		if _, ok := t.postings[key[0]]; !ok {
+			t.postings[key[0]] = map[string]index.Range{}
+		}
+
+		lastKey = key
+		prevRng = index.Range{Start: int64(off + postingLengthFieldSize)}
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "read postings table")
+	}
+
+	if lastKey != nil {
+		prevRng.End = int64(indexLastPostingEnd) - crc32.Size
+		t.postings[lastKey[0]][lastKey[1]] = prevRng
+	}
+
+	return &t, nil
+}
+
+func newV2PostingOffsetTable(factory *stream_encoding.DecbufFactory, tableOffset int, indexLastPostingEnd uint64, postingOffsetsInMemSampling int) (PostingOffsetTable, error) {
+	t := PostingOffsetTableV2{
+		factory:                     factory,
+		tableOffset:                 tableOffset,
+		postings:                    map[string]*postingValueOffsets{},
+		postingOffsetsInMemSampling: postingOffsetsInMemSampling,
+	}
+
+	lastTableOff := 0
+	valueCount := 0
+	var lastKey []string
+
+	// For the postings offset table we keep every label name but only every nth
+	// label value (plus the first and last one), to save memory.
+	if err := readOffsetTable(factory, tableOffset, func(key []string, off uint64, tableOff int) error {
+		if len(key) != 2 {
+			return errors.Errorf("unexpected key length for posting table %d", len(key))
+		}
+
+		if _, ok := t.postings[key[0]]; !ok {
+			// Not seen before label name.
+			t.postings[key[0]] = &postingValueOffsets{}
+			if lastKey != nil {
+				// Always include last value for each label name, unless it was just added in previous iteration based
+				// on valueCount.
+				if (valueCount-1)%postingOffsetsInMemSampling != 0 {
+					t.postings[lastKey[0]].offsets = append(t.postings[lastKey[0]].offsets, postingOffset{value: lastKey[1], tableOff: lastTableOff})
+				}
+				t.postings[lastKey[0]].lastValOffset = int64(off - crc32.Size)
+				lastKey = nil
+			}
+			valueCount = 0
+		}
+
+		lastKey = key
+		lastTableOff = tableOff
+		valueCount++
+
+		if (valueCount-1)%postingOffsetsInMemSampling == 0 {
+			t.postings[key[0]].offsets = append(t.postings[key[0]].offsets, postingOffset{value: key[1], tableOff: tableOff})
+		}
+
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "read postings table")
+	}
+	if lastKey != nil {
+		if (valueCount-1)%postingOffsetsInMemSampling != 0 {
+			// Always include last value for each label name if not included already based on valueCount.
+			t.postings[lastKey[0]].offsets = append(t.postings[lastKey[0]].offsets, postingOffset{value: lastKey[1], tableOff: lastTableOff})
+		}
+		// In any case lastValOffset is unknown as don't have next posting anymore. Guess from TOC table.
+		// In worst case we will overfetch a few bytes.
+		t.postings[lastKey[0]].lastValOffset = int64(indexLastPostingEnd) - crc32.Size
+	}
+	// Trim any extra space in the slices.
+	for k, v := range t.postings {
+		l := make([]postingOffset, len(v.offsets))
```

Comment:
I know this was copied from TSDB code, but we could actually check `cap(t.offsets)` and skip trim if already match the length.

Reply:
Resolved in e14a5b7.